Source/Core/DolphinQt/Config/LogWidget.h

```diff
@@ -43,12 +44,99 @@ class LogWidget final : public QDockWidget, LogListener
   QCheckBox* m_log_wrap;
   QComboBox* m_log_font;
   QPushButton* m_log_clear;
-  QVBoxLayout* m_main_layout;
-  QTextEdit* m_log_text;
-  QWidget* m_tab_log;
+  QPlainTextEdit* m_log_text;
 
   QTimer* m_timer;
 
+  // Ring buffer implementation, heavily based on:
+  // https://embeddedartistry.com/blog/2017/4/6/circular-buffers-in-cc#c-=
+  // Modified to be an almost complete drop-in replacement for std::queue
+  // It does not implement some methods compared to std::queue,
+  // and also implements capacity(), clear() and full(), which std::queue does not
+  template <class T>
+  class circular_buffer
+  {
+  public:
+    using value_type = T;
+    using size_type = std::size_t;
+    using reference = T&;
+    using const_reference = const T&;
+
+    explicit circular_buffer(std::size_t count)
+        : m_container(std::make_unique<storage[]>(count)), m_capacity(count)
+    {
+    }
+
+    bool empty() const noexcept { return m_head == m_tail && !m_full; }
+    bool full() const noexcept { return m_full; }
+    size_type capacity() const noexcept { return m_capacity; }
+
+    size_type size() const noexcept
+    {
+      size_type size;
+      if (m_full)
+        size = m_capacity;
+      else if (m_head >= m_tail)
+        size = m_head - m_tail;
+      else
+        size = m_capacity + m_head - m_tail;
+      return size;
+    }
+
+    const_reference front() const noexcept
+    {
+      return *reinterpret_cast<const_pointer>(std::launder(&m_container[m_tail]));
+    }
+
+    reference front() noexcept
+    {
+      return const_cast<reference>(const_cast<const circular_buffer*>(this)->front());
+    }
+
+    // If buffer is empty, behaviour is undefined
+    void pop() noexcept
+    {
+      reinterpret_cast<pointer>(std::launder(&m_container[m_tail]))->~value_type();
```

Comment:
Note: C++20 marks std::launder as `[[nodiscard]]`. This would discard the return value.

Reply:
Laundered pointer is used for a method call (`~value_type()`) so it's not really discarded and `[[nodiscard]]` doesn't complain about it.